org.eclipse.jdt.text.tests/src/org/eclipse/jdt/text/tests/folding/CustomFoldingRegionNewFoldingTest.java

```diff
@@ -0,0 +1,215 @@
+				package org.example.test;
+				public class Test {
+					void a(){
+						// region
+
+						// endregion
+					}
+				}
+				""";
+		List<IRegion> projectionRanges= getProjectionRangesOfFile(str);
+		assertEquals(2, projectionRanges.size());
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 2, 5);
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 3, 5);
+	}
+
+	@Test
+	@Override
+	public void testCustomFoldingRegionsMultipleLevels() throws Exception {
+		String str= """
+				package org.example.test;
+				// region outside class
+				public class Test {
+					// region outside method
+					void a(){
+						// endregion should be ignored
+						// region inside method
+						System.out.println("Hello World");
+						// endregion inside method
+					}
+					// endregion outside method
+				}
+				// endregion outside class
+				""";
+		List<IRegion> projectionRanges= getProjectionRangesOfFile(str);
+		assertEquals(4, projectionRanges.size());
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 1, 12);//outside class
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 3, 10);//outside method
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 4, 8);//void a()
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 6, 8);//inside method
+	}
+
+	@Test
+	@Override
+	public void testCustomFoldingRegionsUsingSpecialCommentTypes() throws Exception {
+		String str= """
+				package org.example.test;
+
+				public class Test {
+					void a(){
+						/* region multiline
+						*/
+						/** region javadoc */
+						/** endregion javadoc */
+						/* endregion multiline
+						*/
+					}
+				}
+				""";
+		List<IRegion> projectionRanges= getProjectionRangesOfFile(str);
+		assertEquals(5, projectionRanges.size());
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 3, 9);//void a()
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 4, 5);// multiline (comment)
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 4, 9);// multiline (folding region)
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 6, 7);// javadoc
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 8, 9);// multiline (last comment)
+	}
+
+	@Test
+	@Override
+	public void testCustomRegionsWithLocalClass() throws Exception {
+		String str= """
+				package org.example.test;
+
+				public class Test {
+					void a(){
+						// region
+						int i;
+
+						// endregion
+						class Inner{
+
+						}
+					}
+				}
+				""";
+		List<IRegion> projectionRanges= getProjectionRangesOfFile(str);
+		assertEquals(2, projectionRanges.size());
+		FoldingTestUtils.assertContainsRegionUsingStartAndEndLine(projectionRanges, str, 3, 10);//void a()
```

Comment:
Again, I am confused on the end line value.  I assume you will clarify why the end-line matches the endregion comment for custom regions but the end brace line is not part of a method fold.

Reply:
Again, showing the `}` makes sense for folded methods and was added by jakub-suliga's changes.

For custom folding regions, there is no point in showing the `// endregion` comment.

If the behavior of including the `// endregion` line is wanted, I can add that.

Reply:
Ok.  Thanks for the explanation.  It works fine in usage.